Add cluster awareness and decommission docs (#2438)
opensearch-project/documentation-website: _api-reference/cluster-awareness.md

---
layout: default
title: Cluster routing and awareness
nav_order: 16
---

# Cluster routing and awareness

> The Cluster routing and awareness operations assign weights to cluster zones for weighted zonal search requests. Weighted zonal requests ensures that if one of your zones fails, you can spread HTTP and search traffic across zones with lower weights.

Reviewer: ```suggestion
Weights per awareness attribute(say zone) can be assigned to enable weighted search request routing policy. This allows users to control the distribution of search traffic across zones, which can commonly be used for cases like zonal deployments, heterogeneous instance types per zone and weighing away a zone during a zonal failure
```

## HTTP and Path methods

```
PUT /_cluster/routing/awareness/<attribute>/weights
GET /_cluster/routing/awareness/<attribute>/weights?local
GET /_cluster/routing/awareness/<attribute>/weights
```

## Path parameters

Parameter | Type | Description
:--- | :--- | :---
attribute | String | The name of the awareness attribute, usually `zone`. The attribute name must match the values listed in the request body when assigning weights to zones.

## Request body

You can assign weights to zone within the request body of the PUT request. For example, if you have zones with the value `zone_1` and `zone_2`, you can assign weights using a whole number inside the request body.

```
{ 
      "zone_1": "1", 
      "zone_2": "1", 
}
```

When assigned the same weight, the weighted value of the zones are equal. This means that traffic routes to both zones in cases of zonal failure, also called _round robin_ allocation.

## Example: Weighted round robin search

The following sample request creates a round robin shard allocation for search traffic.

### Request

```
PUT /_cluster/routing/awareness/zone/weights
{ 
      "zone_1": "1", 
      "zone_2": "1", 
      "zone_3": "0"
}
```

> 

Reviewer: This has been updated. Please check with @imRishN

### Response

```
{
     "acknowledged": true
}
```

## Example: Get weight for a local node

The following sample request gets the weight for a local node as well as any zones in your cluster. This gives you the ability to see if you local node is weighted over your zones.

### Request

```
GET /_cluster/routing/awareness/zone/weights?local
```

### Response

```json
{ 
    "zone_1": "1", 
    "zone_2": "1", 
    "zone_3": "0",

>     "node_weight" : "0"

Reviewer: This has been updated. Please check with @imRishN

}
```

## Example: Get weights for all zones

The following sample request gets weights for all zones.

### Request

```
GET /_cluster/routing/awareness/<attribute>/weights
```

### Response

OpenSearch responds with the weight of each zone.

```json
{
      "zone_1": "1", 
      "zone_2": "1", 
      "zone_3": "0"
}

> ```

Reviewer: This has been updated. Please check with @imRishN

---
layout: default
title: Cluster routing and awareness
nav_order: 16
---

# Cluster routing and awareness

> To control the distribution of search or HTTP traffic, you can use the weights per awareness attribute to control the distribution of search or HTTP traffic across zones. This is commonly used for zonal deployments, heterogenous instances, and weighing traffic away from zones during zonal failure.

Reviewer 2: Is weighing the right word here? Should it be 'weighting'?

Author: Switched to "routing".

> ## HTTP and Path methods

Reviewer 3: Should path be capitalized here?

```
PUT /_cluster/routing/awareness/<attribute>/weights
GET /_cluster/routing/awareness/<attribute>/weights?local
GET /_cluster/routing/awareness/<attribute>/weights
```

## Path parameters

Parameter | Type | Description
:--- | :--- | :---
attribute | String | The name of the awareness attribute, usually `zone`. The attribute name must match the values listed in the request body when assigning weights to zones.

## Request body parameters

Parameter | Type | Description
:--- | :--- | :---

> weights | JSON object | Assigns weights to attributes within the request body of the PUT request. Weights can be set in any ration, for example, 2:3:5. In a 2:3:5 ratio with three zones, for every 100 requests sent to the cluster, each zone would receive either 20, 30, or 50 search requests in a random order. When assigned a weight of `0`, the zone is cut off from an search traffic. 

Author: Let's switch to "does not receive".

> weights | JSON object | Assigns weights to attributes within the request body of the PUT request. Weights can be set in any ration, for example, 2:3:5. In a 2:3:5 ratio with three zones, for every 100 requests sent to the cluster, each zone would receive either 20, 30, or 50 search requests in a random order. When assigned a weight of `0`, the zone is cut off from an search traffic. 

Reviewer 4: "does not receive any" (remove "from")

_version | String | Implements Optimistic concurrency control (OCC) through versioning. The parameter uses simple versioning, such as `1`, and increments upward based on each subsequent modification. This allows any servers from which a request originates to validate whether or not a zone has been modified. 

In the following example request body, `zone_1` and `zone_2` receive 50 requests each, whereas `zone_3` is cut for from receiving requests.

```
{ 
      "weights":
      {
        "zone_1": "5", 
        "zone_2": "5", 
        "zone_3": "0"
      }
      "_version" : 1
}
```

## Example: Weighted round robin search

> The following sample request creates a round robin shard allocation for search traffic, using an undefined ratio.

Reviewer 2: Nitpicky: sample --> example

Reviewer 2: Delete the comma after traffic?

> The following sample request creates a round robin shard allocation for search traffic, using an undefined ratio.

Reviewer 5: we're using colons to precede code blocks again. "The following example. ..:"

> The following sample request creates a round robin shard allocation for search traffic, using an undefined ratio.

Reviewer 5: ```suggestion
The following sample request creates a round robin shard allocation for search traffic, using an undefined ratio:
```

### Request

PUT /_cluster/routing/awareness/zone/weights
{ 
      "weights":
      {
        "zone_1": "1", 
        "zone_2": "1", 
        "zone_3": "0"
      }
      "_version" : 1
}

### Response

```
{
     "acknowledged": true
}
```

## Example: Get weights for all zones

> The following sample request gets weights for all zones.

Reviewer 2: Nitpicky: sample --> example

Reviewer 2: gets --> returns?

> The following sample request gets weights for all zones.

Reviewer 5: ```suggestion
The following sample request gets weights for all zones:
```
